source/common/upstream/load_balancer_impl.cc

```diff
@@ -238,22 +238,68 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
 
-  // Panic mode is disabled only when panic_threshold is 0%.
-  if (panic_threshold > 0 && normalized_total_availability == 0) {
-    // Everything is terrible. All load should be to P=0. Turn on panic mode.
-    ASSERT(per_priority_load_.healthy_priority_load_.get()[0] == 100);
-    per_priority_panic_[0] = true;
+  // This is corner case when panic is disabled and there is no hosts available.
+  // But the load must be distributed somewhere.
+  // This is just a placeholder, because no traffic will be directed to
+  // priority 0 anyways, because it is empty.
+  if (panic_threshold == 0 && normalized_total_availability == 0) {
+    per_priority_load_.healthy_priority_load_.get()[0] = 100;
     return;
   }
 
+  bool total_panic = true;
   for (size_t i = 0; i < per_priority_health_.get().size(); ++i) {
     // For each level check if it should run in panic mode. Never set panic mode if
     // normalized total health is 100%, even when individual priority level has very low # of
     // healthy hosts.
     const HostSet& priority_host_set = *priority_set_.hostSetsPerPriority()[i];
     per_priority_panic_[i] =
-        (normalized_total_availability == 100 ? false : isGlobalPanic(priority_host_set));
+        (normalized_total_availability == 100 ? false : isHostSetInPanic(priority_host_set));
+    total_panic = total_panic && per_priority_panic_[i];
   }
+
+  // If all priority levels are in panic mode, load distribution
+  // is done differently.
+  if (total_panic) {
+    recalculateLoadInTotalPanic();
+  }
+}
+
+// recalculateLoadInTotalPanic method is called when all priority levels
+// are in panic mode. The load distribution is done NOT based on number
+// of healthy hosts in the priority, but based on number of hosts
+// in each priority regardless of its health.
+void LoadBalancerBase::recalculateLoadInTotalPanic() {
+  // First calculate total number of hosts across all priorities regardless
+  // whether they are healthy or not.
+  const uint32_t total_hosts_num = std::accumulate(
+      priority_set_.hostSetsPerPriority().begin(), priority_set_.hostSetsPerPriority().end(), 0,
+      [](size_t acc, const std::unique_ptr<Envoy::Upstream::HostSet>& host_set) {
+        return acc + host_set->hosts().size();
+      });
+
+  if (0 == total_hosts_num) {
+    // Backend is empty, but load must be distributed somewhere.
+    per_priority_load_.healthy_priority_load_.get()[0] = 100;
+    return;
+  }
+
+  // Now iterate through all priority levels and calculate how much
+  // load is supposed to go to each priority. In panic mode the calculation
+  // is based not on the number of healthy hosts but based on the number of
+  // total hosts in the priority.
+  uint32_t total_load = 100;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    auto hosts_num = host_set.hosts().size() - host_set.excludedHosts().size();
```

Comment:
where did we land on whether to include excluded hosts? if they are used during panic routing i think it would make sense to not subtract them here?

Reply:
Done.

Reply:
I found out that excluded hosts are those for which healthcheck is not resolved yet. I understand that this was added to avoid entering panic mode when bunch of new hosts is added for which health is not known yet. Whether to route to them in panic mode? I guess the answer is yes and that is how it is implemented now.

The existence of `host_set.excludedHosts().size();` here is a bug, because I removed it from calculation of the total load in my previous commit but I did not remove it here when hosts are calculated for each priority. I was perplexed why none of tests caught it. So I added a new test to include excludedHosts case, but the behavior was not what I expected. It happened that there was a bug in test setup routine. The following line is wrong: https://github.com/envoyproxy/envoy/blob/v1.12.2/test/common/upstream/load_balancer_impl_test.cc#L87. It was probably copy-pasted.